File: pt5/1146.snapshot-array.py
from __future__ import annotations

from bisect import bisect_right
from typing import List, Tuple


class SnapshotArray:
    def __init__(self, length: int):
        self.snap_id = 0
        self.history: List[List[Tuple[int, int]]] = [[(0, 0)] for _ in range(length)]

    def set(self, index: int, val: int) -> None:
        if self.history[index][-1][0] == self.snap_id:
            self.history[index][-1] = (self.snap_id, val)
        else:
            self.history[index].append((self.snap_id, val))

    def snap(self) -> int:
        current = self.snap_id
        self.snap_id += 1
        return current

    def get(self, index: int, snap_id: int) -> int:
        records = self.history[index]
        position = bisect_right(records, (snap_id, float("inf"))) - 1
        return records[position][1]

File: pt5/1146.snapshot-array.go
package main

import "sort"

type snapshotValue struct {
	snapID int
	value  int
}

type SnapshotArray struct {
	snapID  int
	history [][]snapshotValue
}

func Constructor(length int) SnapshotArray {
	history := make([][]snapshotValue, length)
	for i := range history {
		history[i] = []snapshotValue{{snapID: 0, value: 0}}
	}
	return SnapshotArray{history: history}
}

func (s *SnapshotArray) Set(index int, val int) {
	records := s.history[index]
	if records[len(records)-1].snapID == s.snapID {
		records[len(records)-1].value = val
		s.history[index] = records
		return
	}
	s.history[index] = append(records, snapshotValue{snapID: s.snapID, value: val})
}

func (s *SnapshotArray) Snap() int {
	current := s.snapID
	s.snapID++
	return current
}

func (s *SnapshotArray) Get(index int, snapID int) int {
	records := s.history[index]
	position := sort.Search(len(records), func(i int) bool {
		return records[i].snapID > snapID
	}) - 1
	return records[position].value
}

